### PR TITLE
GH-36709: [Python] Allow to specify use_threads=False in Table.group_by to have stable ordering

### LINKED ISSUE #36709: [Python] Guarantee that `group_by` has stable ordering.
### Describe the enhancement requested

As far as I can tell, it's not officially documented that grouping maintains the order of its keys, but it has seemed that way until recently. Starting with 13.0.0.dev, I've noticed ordering changes intermittently. I think this is an important feature of grouping, as with sorting.

This example outputs differences reliably on 13.0.0.dev516.
```python
import pyarrow as pa

array = pa.chunked_array([list('aba'), list('bcb')])
table = pa.table({'key': array})
for i in range(100):
    groups = table.group_by(['key']).aggregate([])
    if groups['key'].to_pylist() != list('abc'):
        print(i, groups['key'])
```

### Component(s)

C++, Python

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
### Rationale for this change

Add a `use_threads` keyword to the `group_by` method on Table, and passes this through to the Declaration.to_table call. This also allows to specify `use_threads=False` to get stable ordering of the output, and which is also required to specify for certain aggregations (eg `"first"` will fail with the default of `use_threads=True`)

### Are these changes tested?

Yes, added a test (similar to the one we have for this for `filter`), that would fail (>50% of the times) if the output was no longer ordered.

* Closes: #36709

### CODE REVIEW

**Review by danepitkin — COMMENTED** [review]
LGTM!

**Comment by jorisvandenbossche** [comment]
> I see a failure in the docstring but the parameter seems to be correctly documented

I have no clue to why this is happening or how to fix it. I can reproduce it locally, but :shrug: 
Will look into an ugly hack of skipping this specific method in archery.

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -4599,7 +4599,7 @@ cdef class Table(_Tabular):
         """
         return self.drop_columns(columns)
 
-    def group_by(self, keys):
+    def group_by(self, keys, use_threads=True):
         """Declare a grouping over the columns of the table.
```
Comment (danepitkin) [inline python/pyarrow/table.pxi:4603]: You could try this, I've actually seen this fix "unused parameters" errors in numpydocs before:
```
"""
Declare a grouping over the columns of the table.
...
"""
```

Comment (jorisvandenbossche) [reply python/pyarrow/table.pxi:4603]: Oh .. and that indeed did it :man_shrugging: 
Thanks a lot!

**Pull request merged**